gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/LaunchDagProc.java

```diff
@@ -0,0 +1,216 @@
+    //Add Dag to the map of running dags
+    String dagId = DagManagerUtils.generateDagId(dag).toString();
+    log.info("Initializing Dag {}", DagManagerUtils.getFullyQualifiedDagName(dag));
+    if (dagManagementStateStore.containsDag(dagId)) {
+      log.warn("Already tracking a dag with dagId {}, skipping.", dagId);
+      return;
+    }
+
+    dagManagementStateStore.addDag(dagId, dag);
+    log.debug("Dag {} - determining if any jobs are already running.", DagManagerUtils.getFullyQualifiedDagName(dag));
+
+    //A flag to indicate if the flow is already running.
+    boolean isDagRunning = false;
+    //Are there any jobs already in the running state? This check is for Dags already running
+    //before a leadership change occurs.
+    for (Dag.DagNode<JobExecutionPlan> dagNode : dag.getNodes()) {
+      if (DagManagerUtils.getExecutionStatus(dagNode) == ExecutionStatus.RUNNING) {
+        dagManagementStateStore.addJobState(dagId, dagNode);
+        //Update the running jobs counter.
+        dagManagerMetrics.incrementRunningJobMetrics(dagNode);
+        isDagRunning = true;
+      }
+    }
+
+    FlowId flowId = DagManagerUtils.getFlowId(dag);
+    this.dagManagerMetrics.registerFlowMetric(flowId, dag);
+
+    log.debug("Dag {} submitting jobs ready for execution.", DagManagerUtils.getFullyQualifiedDagName(dag));
+    //Determine the next set of jobs to run and submit them for execution
+    Map<String, Set<Dag.DagNode<JobExecutionPlan>>> nextSubmitted = submitNext(dagManagementStateStore, dagId);
+    for (Dag.DagNode<JobExecutionPlan> dagNode : nextSubmitted.get(dagId)) {
+      dagManagementStateStore.addJobState(dagId, dagNode);
+    }
+
+    // Set flow status to running
+    DagManagerUtils.emitFlowEvent(Optional.of(this.eventSubmitter), dag, TimingEvent.FlowTimings.FLOW_RUNNING);
+    dagManagerMetrics.conditionallyMarkFlowAsState(flowId, DagManager.FlowState.RUNNING);
+
+    // Report the orchestration delay the first time the Dag is initialized. Orchestration delay is defined as
+    // the time difference between the instant when a flow first transitions to the running state and the instant
+    // when the flow is submitted to Gobblin service.
+    if (!isDagRunning) {
+      this.orchestrationDelay.set(System.currentTimeMillis() - DagManagerUtils.getFlowExecId(dag));
+    }
+
+    log.info("Dag {} Initialization complete.", DagManagerUtils.getFullyQualifiedDagName(dag));
+  }
+
+  /**
+   * Submit next set of Dag nodes in the Dag identified by the provided dagId
+   * @param dagId The dagId that should be processed.
+   * @return
+   * @throws IOException
+   */
+  // todo - convert to return set only
+  synchronized Map<String, Set<Dag.DagNode<JobExecutionPlan>>> submitNext(DagManagementStateStore dagManagementStateStore, String dagId) {
+    Dag<JobExecutionPlan> dag = dagManagementStateStore.getDag(dagId);
+    Set<Dag.DagNode<JobExecutionPlan>> nextNodes = DagManagerUtils.getNext(dag);
```

Comment:
can getNext return null? let's be safe in case it's called on a dag with no remaining nodes

Reply:
No, getNext never returns a null